### ISSUE TITLE
SIGILL in _armv7_tick on ARM when using openssl 1.1.1k

### ISSUE THREAD

**Issue by mahendra-msp** [openssl/openssl#17465]
Hi,

We are using OpenSSL 1.1.1k.  Intermittently we see SIGILL crash in _armv7_tick on ARM platform in the RAND_DRBG_bytes path as below.
#4  <signal handler called>
#5  _armv7_tick () at crypto/armv4cpuid.S:136
#6  0xf57a00e8 in get_timer_bits () at ../openssl-1.1.1k/crypto/rand/rand_unix.c:833
#7  rand_pool_add_additional_data (pool=pool@entry=0xf1c5cd80) at ../openssl-1.1.1k/crypto/rand/rand_unix.c:789
#8  0xf579f46e in rand_drbg_get_additional_data (pool=0xf1c5cd80, pout=pout@entry=0xc8afe1e0) at ../openssl-1.1.1k/crypto/rand/rand_lib.c:289
#9  0xf579ea8e in RAND_DRBG_bytes (drbg=0xf1c6ef50, out=0xc8afe260 "", outlen=4) at ../openssl-1.1.1k/crypto/rand/drbg_lib.c:671

Found similar issues here (https://github.com/openssl/openssl/issues/14838, https://github.com/openssl/openssl/issues/17009). 
Could you please help on how we can resolve this issue ?

Thanks
Mahendra

**Comment by bernd-edlinger** [comment]
It is interesting that this did not happen when the feature is first probed on startup.
You could set an environment to prevent this:
`export OPENSSL_armcap=0`
the bits in the armcap are defined like this:
```
# define ARMV7_NEON      (1<<0)
# define ARMV7_TICK      (1<<1)
# define ARMV8_AES       (1<<2)
# define ARMV8_SHA1      (1<<3)
# define ARMV8_SHA256    (1<<4)
# define ARMV8_PMULL     (1<<5)
# define ARMV8_SHA512    (1<<6)
```
so 0 should be safe, but maybe result in less performance.

**Comment by mahendra-msp** [comment]
@bernd-edlinger  Thank you for the feedback. 
I see that here https://github.com/catboost/catboost/commit/d7c6f1d7388ee9a70c2f26d0e388f4d83f61b5b5, setting ARMV7_TICK is avoided. Is this an alternate way to explore? Is there any way to avoid impact on performance?

Thanks
Mahendra

**Comment by bernd-edlinger** [comment]
when you encountered that problem, are you sure, that the OPENSSL_armcap was not set ?

**Comment by dlegaultbbry** [comment]
This was reported previously if you have a system which uses Big.little where one set of CPUs can support some instructions and others not. See https://github.com/openssl/openssl/issues/14838

**Comment by bernd-edlinger** [comment]
@mahendra-msp do you have such a non-symmetric-multi-processor-system, where some cores do support different instruction sets?

**Comment by paulidale** [comment]
Currently we only access the timer for two purposes:

1. additional data for the RNG
2. seeding for the stochastic property cache flush

Neither of these needs to be fast, so we could add a dynamic check in the two armcap files or wrapper the call in armcap.c with an illegal instruction catcher.

**Comment by bernd-edlinger** [comment]
> add a dynamic check in the two armcap files or wrapper the call in armcap.c with an illegal instruction catcher.

Sorry, but I don't think that is a good idea, since changing the signal handlers affects the whole process, and is
therefore only acceptable because we do it only once when all the static constructors run (for ARM at least).

**Comment by bernd-edlinger** [comment]
> Is there any way to avoid impact on performance?

You could of course also use my feature branch at #11900 
and set `OPENSSL_armcap=0x3d` which makes it probe all features except ARMV7_TICK.

**Comment by paulidale** [comment]
Setting `OPENSSL_armcap` to the features that are supported on your device should avoid this problem without impacting performance.  The timer isn't used for anything critical.

**Comment by mahendra-msp** [comment]
Thank you all for the inputs.

Confirmed that OPENSSL_armcap_P value is not read from environment variable. It is set the end of OPENSSL_cpuid_setup routine to the value 611 (0x263). This value does not get modified during the course of firmware execution.

@paulidale do you mean, we can reset ARMV7_TICK bit in OPENSSL_armcap_P and keep the rest of the bits unmodified to address this issue?

**Comment by bernd-edlinger** [comment]
> It is set the end of OPENSSL_cpuid_setup routine to the value 611 (0x263)

I don't understand what the bit 0x200 bit means.
Do you use 1.1.1m or some other version?

**Comment by mahendra-msp** [comment]
@bernd-edlinger  I am re-checking OPENSSL_armcap_P value. I will confirm here again. We are using openssl 1.1.1k

**Comment by paulidale** [comment]
> @paulidale do you mean, we can reset ARMV7_TICK bit in OPENSSL_armcap_P and keep the rest of the bits unmodified to address this issue?

You should be able to set the `OPENSSL_armcap` environment variable to work around this.  If this environment variable is defined, it's value is used instead of the probing.  You need to match it to your system sans the ARMV7_TICK bit.

**Comment by mahendra-msp** [comment]
@bernd-edlinger I could re-check the OPENSSL_armcap_P and is set to 63 (0x3F). So all bits are set in OPENSSL_cpuid_setup()

@bernd-edlinger  @paulidale Can you please give more info on the reason for this SIGILL error ? It would help us to know the issue better.

**Comment by paulidale** [comment]
You could answer the [question above](https://github.com/openssl/openssl/issues/17465#issuecomment-1010367827) for starters.

Your processor claims to have the ARMV7_TICK register but then fails to access it.

**Comment by bernd-edlinger** [comment]
> the OPENSSL_armcap_P and is set to 63 (0x3F).

Okay, then please add this to your environment: `OPENSSL_armcap=0x3D`.

> Can you please give more info on the reason for this SIGILL error ? It would help us to know the issue better.

Other users did report they have observed something like this, because they
used a multi core system with several arm cores of **different** type.
Only during the initial probing a SIGILL exception is caught by openssl, but we do
not track that feature per-core.

**Comment by mahendra-msp** [comment]
@bernd-edlinger @paulidale Thank you so much for the above inputs and suggestions. I will get back in case of any further queries. Thank you.

**Comment by marler8997** [comment]
I may have found an issue with this code that may or may not be related.

From the stack trace, it looks like we get the a signal on this `MRRC` instruction:
```
	mrrc	p15,1,r0,r1,c14		@ CNTVCT
```

Make note that the C prototype for this function is defined as `uint32_t _armv7_tick(void)`, and the assembly code is just 2 instructions, the `mrrc` one shown above and then a `bx lr` to return.

The MRRC instruction is taking the 64-bit coprocessor register `c14` and storing it in 2 32-bit registers `r0` and `r1`.  I'm not sure what the C ABI is defined for arm32, but my guess is 1 of these registers is used for the return value.  The question is, in the C ABI is the caller responsible for restoring all the other registers?  If so then it's possible this `armv7_tick` function is clobbering one of these registers and the compiler doesn't know about it.

**Comment by davidben** [comment]
That does not appear to be related, or actually an issue.

> The question is, in the C ABI is the caller responsible for restoring all the other registers? If so then it's possible this armv7_tick function is clobbering one of these registers and the compiler doesn't know about it.

I think you may have inverted your first question, but, no, clobbering those registers is fine. As the calling convention says, r0-r7 can only be used to hold intermediate values in between subroutine calls.
https://github.com/ARM-software/abi-aa/blob/main/aapcs64/aapcs64.rst#general-purpose-registers

Which makes sense. If your parameter-passing registers are non-volatile, every non-leaf function needs to save and restore them defensively, which would be wasted saves and restores most of the time.

(Even if there were a calling convention mistake, it would likely manifest as a more arbitrary crash or misbehavior, not a SIGILL.)

> I wrote a small test app and when I enabled full-optimizations, if I didn't tell the inline-assembly that both registers were being clobbered in some way then it would result in unpredictable behavior.

Inline assembly isn't a function call. Compilers typically require inline assembly to describe all registers clobbered. Not doing so is a bug in your inline assembly snippet and has no connection to the calling convention.
https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html

**Comment by marler8997** [comment]
Yeah I can confirm pretty much everything you said @davidben.  So my best guess as to what's causing this now is maybe this instruction can cause an invalid instruction interrupt at certain times, maybe when the processor is in a particular power state?  I'm trying to write a test app to reproduce this but have failed to do so this far, even with 5 threads going at it in tight loops.

**Comment by jarkkokoivikko-code-q** [comment]
I think this is the same problem with my application. It happens irregularly. I would be happy to try a fix if one is available.

<pre>
: Fatal signal 4 (SIGILL), code 1 (ILL_ILLOPC), fault addr 0xb75ec9a8 in tid 25156 (qtMainLoopThrea), pid 25405 (x)
: *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
: Build fingerprint: 'samsung/gta4xlwifieea/gta4xlwifi:11/RP1A.200720.012/x:user/release-keys'
: Revision: '7'
: ABI: 'arm'
: Processor: '4'
: Timestamp: 2022-01-19 10:21:11+0200
: pid: 25405, tid: 25156, name: qtMainLoopThrea  >>> x <<<
: uid: 10327
: signal 4 (SIGILL), code 1 (ILL_ILLOPC), fault addr 0xb75ec9a8 (*pc=0xec510f1e)
:     r0  0000003f  r1  b98e813c  r2  00000030  r3  b78a3050
:     r4  b77d6418  r5  00000000  r6  c7112eb8  r7  00000000
:     r8  b7be5a00  r9  b3eb4cc0  r10 b98e82e0  r11 b98e8118
:     ip  b77b4fc4  sp  b98e80f8  lr  b76fdc7c  pc  b75ec9a8
: backtrace:
:       #00 pc 000579a8  /data/app/~~C8YCYg4LdUM3oG9I937aiw==/x-A6levaWv78H5Qvs3p44u-A==/lib/arm/libcrypto_1_1.so
</pre>

**Comment by davidben** [comment]
See #14838 which already includes a diagnosis of this problem for big.little systems. I suspect other folks running into this are running into the same thing.

The probing stuff OpenSSL does isn't a safe way to detect ARM features and should be removed in favor of supported APIs. Unfortunately, I couldn't find a relevant hwcap value for this instruction, so I guess the question is what is OpenSSL doing with it. Perhaps you all can find another way to do it... I see some entropy bits, but there are better sources of entropy, and something in code new to OpenSSL 3.0 to do something odd deep in the property system.

**Comment by marler8997** [comment]
Our system isn't big little. We use a cortex A72 which has 4 identical cores. Furthermore, if this was the issue my test app running 5 threads calling the instruction as possible should have reproduced it almost instantly. We do alot of "duration testing" and have only seen this specific problem happen about 2 or 3 times after the device has been running for days going in and out of sleep.

**Comment by bernd-edlinger** [comment]
That sounds almost like an A72 erratum.
Have you informed ARM about this?

**Comment by marler8997** [comment]
Will do if I can find a way to reliably reproduce it.

**Comment by Superpengun** [comment]
From the previous records, the cause of this problem still needs to be investigated............

**Comment by love-ginger** [comment]
I encountered the same issue，Openssl 1.1.1l，OPPO R15 PACM00，Android 10

**Comment by t8m** [comment]
The version 1.1.1 is not supported anymore. Closing. If this is still an issue with currently supported releases, please open a new issue.
